wip/6.0 HHH-14091: Replace bunch of if-else-statement with enum

yangchenwen opened the pull request:
Hi Steve @sebersole , perhaps we don't have to worry about the performance loss this time, because I just refactored this change using a enum chain instead of a enum loop. Please take a look.

NathanQingyangXu commented:
Do we have testing code covered for the code changes? If no legacy testing cases exist, it would be great to create complete testing code to showcase your code change effectiveness.
From coding perspective, it seems great clearing up.
Usually we adopt the PR name pattern of 'wip/6.0 xxxx' to make the PR stand out among ones targeting v5.

yangchenwen commented:
> Do we have testing code covered for the code changes? If no legacy testing cases exist, it would be great to create complete testing code to showcase your code change effectiveness.
> From coding perspective, it seems great clearing up.
> Usually we adopt the PR name pattern of 'wip/6.0 xxxx' to make the PR stand out among ones targeting v5.

Actually there's a legacy testing case in org.hibernate.jdbc.util.BasicFormatterTest. I didn't add any new feature to the formatter,  so the legacy case should cover the changes. I already rename the PR by adding a 'wip/6.0' prefix.  Thanks for reminding!

sebersole commented:
Actually the best way to denote a PR as being against 6.0 is to apply the
label `6.0`.  Personally, I look for that label when looking at 6.0 PRs.

FWIW I always call my PR branches `topic/xxx` or `pr/xxx`.

On Wed, Jul 1, 2020 at 3:18 PM Nathan Xu <notifications@github.com> wrote:

> Do we have testing code covered for the code changes? If no legacy testing
> cases exist, it would be great to create complete testing code to showcase
> your code change effectiveness.
> From coding perspective, it seems great clearing up.
> Usually we adopt the PR name pattern of 'wip/6.0 xxxx' to make the PR
> stand out among ones targeting v5.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/hibernate/hibernate-orm/pull/3446#issuecomment-652625705>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABZIE7FIQC5DJWK22LYAH3RZOKYNANCNFSM4ONIGGSQ>
> .
>

NathanQingyangXu commented on hibernate-core/src/main/java/org/hibernate/engine/jdbc/internal/BasicFormatterImpl.java:
```diff
@@ -110,127 +110,48 @@ public String perform() {
+						break;
+					case "\"": {
+						concatUntil( "\"" );
+						break;
```
could we combine the above two cases as following:
```
switch ( token ) {
    case "'":
    case "\"":
         concatUntil( token );
         break;
     ... ...
}
```

NathanQingyangXu commented on hibernate-core/src/main/java/org/hibernate/engine/jdbc/internal/BasicFormatterImpl.java:
```diff
@@ -390,6 +201,261 @@ private void newline() {
+			PROCESS_AFTER_BETWEEN_AND {
+				@Override
+				public boolean process(FormatProcess p) {
+					if ( p.afterBetween && p.lcToken.equals( "and" ) ) {
```
Using `"and".equals( p.lcToken )` would be more consistent and robust, IMHO.

yangchenwen replied:
Yeah you're right, I'll fix it later.

NathanQingyangXu commented:
> Actually the best way to denote a PR as being against 6.0 is to apply the label `6.0`. Personally, I look for that label when looking at 6.0 PRs. FWIW I always call my PR branches `topic/xxx` or `pr/xxx`.
> […](#)
> On Wed, Jul 1, 2020 at 3:18 PM Nathan Xu ***@***.***> wrote: Do we have testing code covered for the code changes? If no legacy testing cases exist, it would be great to create complete testing code to showcase your code change effectiveness. From coding perspective, it seems great clearing up. Usually we adopt the PR name pattern of 'wip/6.0 xxxx' to make the PR stand out among ones targeting v5. — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub <[#3446 (comment)](https://github.com/hibernate/hibernate-orm/pull/3446#issuecomment-652625705)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AABZIE7FIQC5DJWK22LYAH3RZOKYNANCNFSM4ONIGGSQ> .

Unluckily, community contributors have no privilege to add label. Each time when my PR whose name starts with 'wip/6.0' is created, it is @dreab8 who added the `6.0` label for me. So much thanks to him.

yangchenwen commented:
> Looking at this some more, you changed this actually to do a loop (enums) within the loop (while). That seems like a not great idea from performance perspective. I think we ought to performance test this change before accepting.

You sure right. Please do the performance test. Also I just moved `ProcessEnum[] processEnums = ProcessEnum.values();` outside the while loop,  it may help a little.

yangchenwen commented:
> > Looking at this some more, you changed this actually to do a loop (enums) within the loop (while). That seems like a not great idea from performance perspective. I think we ought to performance test this change before accepting.
> 
> You sure right. Please do the performance test. Also I just moved `ProcessEnum[] processEnums = ProcessEnum.values();` outside the while loop. I think it may help a little.

NathanQingyangXu commented:
> > Looking at this some more, you changed this actually to do a loop (enums) within the loop (while). That seems like a not great idea from performance perspective. I think we ought to performance test this change before accepting.
> 
> You sure right. Please do the performance test. Also I just moved `ProcessEnum[] processEnums = ProcessEnum.values();` outside the while loop. I think it may help a little.

I think Steve's suggestion is your conducting the performance test (e.g. JMH comparison between the old code and your new code. see https://www.baeldung.com/java-microbenchmark-harness). I can help you if you want (you can send private chat message on zulip: https://hibernate.zulipchat.com/#narrow/stream/132094-hibernate-orm-dev, btw).

yangchenwen commented:
> > > Looking at this some more, you changed this actually to do a loop (enums) within the loop (while). That seems like a not great idea from performance perspective. I think we ought to performance test this change before accepting.
> > 
> > 
> > You sure right. Please do the performance test. Also I just moved `ProcessEnum[] processEnums = ProcessEnum.values();` outside the while loop. I think it may help a little.
> 
> I think Steve's suggestion is your conducting the performance test (e.g. JMH comparison between the old code and your new code. see https://www.baeldung.com/java-microbenchmark-harness). I can help you if you want (you can send private chat message on zulip: https://hibernate.zulipchat.com/#narrow/stream/132094-hibernate-orm-dev, btw).



> > > Looking at this some more, you changed this actually to do a loop (enums) within the loop (while). That seems like a not great idea from performance perspective. I think we ought to performance test this change before accepting.
> > 
> > 
> > You sure right. Please do the performance test. Also I just moved `ProcessEnum[] processEnums = ProcessEnum.values();` outside the while loop. I think it may help a little.
> 
> I think Steve's suggestion is your conducting the performance test (e.g. JMH comparison between the old code and your new code. see https://www.baeldung.com/java-microbenchmark-harness). I can help you if you want (you can send private chat message on zulip: https://hibernate.zulipchat.com/#narrow/stream/132094-hibernate-orm-dev, btw).

Oh that's awkward! I thought "we" Steve said meant the hibernate team :) 
Anyway, I'll do the performance test. I just wondering if there is any rule I should follow or any requirement I should meet while writing the test code, and how to present the test result, because I can't find any referable performance test among the test cases. 
Thanks for your offering to help, Nathan. It's very kind of you.

NathanQingyangXu commented:
There is no performance testing cases in Hibernate I think. Guess you only
need to present the benchmark comparison results in some comment.

On Tue., Aug. 4, 2020, 07:15 yangchenwen, <notifications@github.com> wrote:

> Looking at this some more, you changed this actually to do a loop (enums)
> within the loop (while). That seems like a not great idea from performance
> perspective. I think we ought to performance test this change before
> accepting.
>
> You sure right. Please do the performance test. Also I just moved ProcessEnum[]
> processEnums = ProcessEnum.values(); outside the while loop. I think it
> may help a little.
>
> I think Steve's suggestion is your conducting the performance test (e.g.
> JMH comparison between the old code and your new code. see
> https://www.baeldung.com/java-microbenchmark-harness). I can help you if
> you want (you can send private chat message on zulip:
> https://hibernate.zulipchat.com/#narrow/stream/132094-hibernate-orm-dev,
> btw).
>
> Looking at this some more, you changed this actually to do a loop (enums)
> within the loop (while). That seems like a not great idea from performance
> perspective. I think we ought to performance test this change before
> accepting.
>
> You sure right. Please do the performance test. Also I just moved ProcessEnum[]
> processEnums = ProcessEnum.values(); outside the while loop. I think it
> may help a little.
>
> I think Steve's suggestion is your conducting the performance test (e.g.
> JMH comparison between the old code and your new code. see
> https://www.baeldung.com/java-microbenchmark-harness). I can help you if
> you want (you can send private chat message on zulip:
> https://hibernate.zulipchat.com/#narrow/stream/132094-hibernate-orm-dev,
> btw).
>
> Oh that's awkward! I thought "we" Steve said meant the hibernate team :)
> Anyway, I'll do the performance test. I just wondering if there is any
> rule I should follow or any requirement I should meet while writing the
> test code, and how to present the test result, because I can't find any
> referable performance test among the test cases.
> Thanks for your offering to help, Nathan. It's very kind of you.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/hibernate/hibernate-orm/pull/3446#issuecomment-668535443>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB6UYASUU32R6GWXUCPNXH3R67UWPANCNFSM4ONIGGSQ>
> .
>

yangchenwen commented:
> There is no performance testing cases in Hibernate I think. Guess you only need to present the benchmark comparison results in some comment.
> […](#)
> On Tue., Aug. 4, 2020, 07:15 yangchenwen, ***@***.***> wrote: Looking at this some more, you changed this actually to do a loop (enums) within the loop (while). That seems like a not great idea from performance perspective. I think we ought to performance test this change before accepting. You sure right. Please do the performance test. Also I just moved ProcessEnum[] processEnums = ProcessEnum.values(); outside the while loop. I think it may help a little. I think Steve's suggestion is your conducting the performance test (e.g. JMH comparison between the old code and your new code. see https://www.baeldung.com/java-microbenchmark-harness). I can help you if you want (you can send private chat message on zulip: https://hibernate.zulipchat.com/#narrow/stream/132094-hibernate-orm-dev, btw). Looking at this some more, you changed this actually to do a loop (enums) within the loop (while). That seems like a not great idea from performance perspective. I think we ought to performance test this change before accepting. You sure right. Please do the performance test. Also I just moved ProcessEnum[] processEnums = ProcessEnum.values(); outside the while loop. I think it may help a little. I think Steve's suggestion is your conducting the performance test (e.g. JMH comparison between the old code and your new code. see https://www.baeldung.com/java-microbenchmark-harness). I can help you if you want (you can send private chat message on zulip: https://hibernate.zulipchat.com/#narrow/stream/132094-hibernate-orm-dev, btw). Oh that's awkward! I thought "we" Steve said meant the hibernate team :) Anyway, I'll do the performance test. I just wondering if there is any rule I should follow or any requirement I should meet while writing the test code, and how to present the test result, because I can't find any referable performance test among the test cases. Thanks for your offering to help, Nathan. It's very kind of you. — You are receiving this because you commented. Reply to this email directly, view it on GitHub <[#3446 (comment)](https://github.com/hibernate/hibernate-orm/pull/3446#issuecomment-668535443)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB6UYASUU32R6GWXUCPNXH3R67UWPANCNFSM4ONIGGSQ> .

OK. I see. thank you!

sebersole commented:
We use JMH for some benchmark testing, but it is not easy (IMO) to include
in an upstream project - so we generally do it in a separate project that
consumes (depends on) the version(s) of Hibernate we want to test.  JMH can
be a bit tricky to set up

On Tue, Aug 4, 2020 at 7:10 AM yangchenwen <notifications@github.com> wrote:

> There is no performance testing cases in Hibernate I think. Guess you only
> need to present the benchmark comparison results in some comment.
> … <#m_-8895759240912573054_>
> On Tue., Aug. 4, 2020, 07:15 yangchenwen, *@*.***> wrote: Looking at this
> some more, you changed this actually to do a loop (enums) within the loop
> (while). That seems like a not great idea from performance perspective. I
> think we ought to performance test this change before accepting. You sure
> right. Please do the performance test. Also I just moved ProcessEnum[]
> processEnums = ProcessEnum.values(); outside the while loop. I think it may
> help a little. I think Steve's suggestion is your conducting the
> performance test (e.g. JMH comparison between the old code and your new
> code. see https://www.baeldung.com/java-microbenchmark-harness). I can
> help you if you want (you can send private chat message on zulip:
> https://hibernate.zulipchat.com/#narrow/stream/132094-hibernate-orm-dev,
> btw). Looking at this some more, you changed this actually to do a loop
> (enums) within the loop (while). That seems like a not great idea from
> performance perspective. I think we ought to performance test this change
> before accepting. You sure right. Please do the performance test. Also I
> just moved ProcessEnum[] processEnums = ProcessEnum.values(); outside the
> while loop. I think it may help a little. I think Steve's suggestion is
> your conducting the performance test (e.g. JMH comparison between the old
> code and your new code. see
> https://www.baeldung.com/java-microbenchmark-harness). I can help you if
> you want (you can send private chat message on zulip:
> https://hibernate.zulipchat.com/#narrow/stream/132094-hibernate-orm-dev,
> btw). Oh that's awkward! I thought "we" Steve said meant the hibernate team
> :) Anyway, I'll do the performance test. I just wondering if there is any
> rule I should follow or any requirement I should meet while writing the
> test code, and how to present the test result, because I can't find any
> referable performance test among the test cases. Thanks for your offering
> to help, Nathan. It's very kind of you. — You are receiving this because
> you commented. Reply to this email directly, view it on GitHub <#3446
> (comment)
> <https://github.com/hibernate/hibernate-orm/pull/3446#issuecomment-668535443>>,
> or unsubscribe
> https://github.com/notifications/unsubscribe-auth/AB6UYASUU32R6GWXUCPNXH3R67UWPANCNFSM4ONIGGSQ
> .
>
> OK. I see. thank you!
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/hibernate/hibernate-orm/pull/3446#issuecomment-668557163>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABZIE43NXMUMCF7LUOOUZTR673C5ANCNFSM4ONIGGSQ>
> .
>

NathanQingyangXu commented:
maybe you can rerun the JMH test to justify the enum chain's benefit?

yangchenwen commented:
> maybe you can rerun the JMH test to justify the enum chain's benefit?

Sure. Please see the below result.

---
>`testFormat_new_once` uses the formatter with enums chain(call once)  
>`testFormat_old_once` uses the original formatter(call once) 

 Run complete. Total time: 00:00:44

|Benchmark  |  Mode | Cnt |  Score  |  Error | Units|
| :-------- | :---  | :---| :----   | :----- | ----: |
|Main.testFormat_new_once | thrpt |   5 |  0.015 ± 0.001 |  |ops/us|
|Main.testFormat_old_once | thrpt |   5 |  0.015 ± 0.001 |  |ops/us|
|Main.testFormat_new_once |  avgt |   5 | 63.081 ± 2.136 |  | us/op|
|Main.testFormat_old_once |  avgt |   5 | 65.709 ± 0.564 |  | us/op|

As we can see, it's pretty much the same.

More results : https://github.com/yangchenwen/benchmark/blob/master/README.md
Benchmark test case source code: https://github.com/yangchenwen/benchmark/blob/master/src/main/java/com/chiliyum/benchmark/Main.java

NathanQingyangXu commented on hibernate-core/src/main/java/org/hibernate/engine/jdbc/internal/BasicFormatterImpl.java:
```diff
@@ -390,6 +196,268 @@ private void newline() {
+				@Override
+				public void process(FormatProcess p) {
+					if ( "(".equals( p.token ) ) {
+						if ( isFunctionName( p.lastToken ) || p.inFunction > 0 ) {
```
Super minor. It might be a good practice to swap the operators around `||` to boost performance in case of `p.inFunction > 0` for it seems `isFunctionName( p.lastToken )` is more expensive ones than `p.inFunction > 0`:
```
if ( p.inFunction > 0 || isFunctionName( p.lastToken ) ) {
    ... ...
}
```
The rationale is when `p.inFunction > 0`, the more expensive `isFunctionName( p.lastToken )` would be skipped due to short circuiting.

Feel free to ignore this comment though.

yangchenwen replied:
IMHO, unless we can predict that `p.inFunction > 0` being true happens more often (in fact the sqls to be formatted can be very random, plus, `isFunctionName( p.lastToken )` being true is the reason of  `p.inFunction > 0` being true), otherwise  the only thing I can do is hoping that the JIT compiler can do something about it :)

NathanQingyangXu commented on hibernate-core/src/main/java/org/hibernate/engine/jdbc/internal/BasicFormatterImpl.java:
```diff
@@ -110,127 +110,43 @@ public String perform() {
 
```
It might be an idea to put `Enum` to further use to improve the various `Set` or `Map` looking up. Take `BEGIN_CLAUSES` as an example. We might first define its own enum class as:
```
private enum BeginClause {
    ... ...
}
```
Then replace `Set` or `Map` with `EnumMap` which is compact and cannot be beat in terms of performance for basically it is array based. The idea might apply for we have many such small-sized enum candidates for the static constants.
Not 100% sure of my comment's validity though.

Sanne commented:
There are some good ideas in here, such as using a `switch` rather than a series of if statements.
But I really doubt that using the enums this ways is more efficient - it's intentionally creating megamorphic code, which is never a good thing.

Why not use method handles if you need different "actions" to process?

Also, I would suggest to propose such PRs organizing them in smaller commit steps; for example if the work which converts the if statements into a switch was done in a first, clean commit, I think we'd have already merged that.

yangchenwen commented:
> There are some good ideas in here, such as using a `switch` rather than a series of if statements.
> But I really doubt that using the enums this ways is more efficient - it's intentionally creating megamorphic code, which is never a good thing.
> 
> Why not use method handles if you need different "actions" to process?
> 
> Also, I would suggest to propose such PRs organizing them in smaller commit steps; for example if the work which converts the if statements into a switch was done in a first, clean commit, I think we'd have already merged that.

I see your point, though method handles have almost the same performance as enums do (if I didn't do wrong about the benchmark tests  **[here](https://github.com/yangchenwen/benchmark)**).
I'll create another PR sometime once I finished refactoring using method handles.  Thanks for the advice!
